### ISSUE TITLE
[BUG] DataFrame.join(Series) does not maintain order of Series values in mode.pandas_compatible

### ISSUE THREAD

**Issue by mroeschke** [rapidsai/cudf#14001]
**Describe the bug**
DataFrame.join(Series) does not maintain order of Series values

**Steps/Code to reproduce bug**
```python
In [12]: import cudf

In [13]: df = cudf.DataFrame([1, 2])

In [16]: ser = cudf.Series(list("abcdef"), index=[0, 0, 0, 1, 1, 1], name="var1")

In [17]: df.join(ser)
Out[17]: 
   0 var1
0  1    a
0  1    c
0  1    b
1  2    d
1  2    e
1  2    f

In [18]: df.to_pandas().join(ser.to_pandas())
Out[18]: 
   0 var1
0  1    a
0  1    b
0  1    c
1  2    d
1  2    e
1  2    f
```

**Expected behavior**
I would expect the order of the joined values to be maintained.

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: conda
   - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**
Please run and paste the output of the `cudf/print_env.sh` script here, to gather any other relevant environment details

**Additional context**

**Comment by bdice** [comment]
We should only expect this is a bug when `mode.pandas_compatible` is set to `True`, right?

Otherwise these docs about [no result ordering guarantee](https://docs.rapids.ai/api/cudf/stable/user_guide/pandas-comparison/#result-ordering) applies, from what I can tell.

**Comment by mroeschke** [comment]
Ah totally forgot about the ordering guarantee. Yes, this is probably a bug only when `cudf.set_option("mode.pandas_compatible", True)` (the OP still reproduces even when setting it)

**Comment by bdice** [comment]
This will likely require some nontrivial changes, like adding a sequential index column before joining, then sorting after the join? Or perhaps some intermediate sorting of the join indices (can't recall the implementation details to know if this is viable at the moment). Either way, preserving order for a hash join is not ideal for performance, which is why this hasn't been done yet. I am not sure if there's any better way -- we did find some clever ways to implement `drop_duplicates` in an order-preserving way, and perhaps the same can be done here. I suspect libcudf changes will be needed to resolve this (at least, in an efficient way). If you'd like me to pursue this, please let me know. Some related issues/PRs:
-  #12236
- #1781
- #11656

**Comment by mroeschke** [comment]
Relative to the xdf effort, I would say this is lower priority to fix

**Comment by shwina** [comment]
> like adding a sequential index column before joining, then sorting after the join

I think we should go ahead and implement this solution for now (when Pandas compatibility is enabled).

**Comment by shwina** [comment]
@mroeschke -- by default, `sort=False` in the `.join()` method.

Is it an implementation detail of Pandas that when `sort=False`, ordering of keys is preserved? Or is it a guarantee? (I couldn't find where in the docs this is defined).

cc: @wence-

**Comment by mroeschke** [comment]
`join` uses `merge` under the hood so the key ordering follow what's described in the `how` section here: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html

**Comment by wence-** [comment]
Cleaning up assigned issues, this is now fixed in 24.12, sorry for the delay!
